src/content/plugins.md

```diff
@@ -0,0 +1,325 @@
+transforms the array of files into an array of the contents of each file. We then join all the files together
+with the separator character, which is returned as the result to the reduce method above. 
+
+After the above is complete, we now have all of the file contents within the `content` variable and all that is 
+left is to write that to the outputFile. 
+
+That's it, plugin complete. In our case, the plugin is entirely synchronous and as such just returns at the end.
+If we needed the plugin to be asynchronous, we could alternatively return a promise and Broccoli would wait until
+the promise resolved before continuing the build.
+
+Lastly, we export a default function that will be used when this plugin is imported into a build pipeline
+```js
+export default function concatPlugin(...params) {
+    return new ConcatPlugin(...params);
+}
+```
+
+E.g.
+
+```js
+// Brocfile.js
+import concat from 'concat-plugin';
+
+export default () => concat(['dir1', 'dir2']);
+```
+
+As you can see, there isn't really any magic happening here, it's all standard Node code, just wrapped in a 
+`build()` method that is provided an array of `inputPaths` and an `outputPath` to write to.
+
+# Caching
+
+Adding caching to a plugin is relatively straight forward. Caching allows a plugin to skip its `build()` method
+if it doesn't need to do anything because its inputs have no changed. We can use a couple of node packages to
+achieve this. [fs-tree-diff](https://www.npmjs.com/package/fs-tree-diff) is used to calculate a diff between the
+2 states of a directory, and [walk-sync](https://www.npmjs.com/package/walk-sync) which synchronously walks all
+the files in a directory and produces a list, which is passed to `fs-tree-diff`.
+
+Let's take a look:
+
+```js
+import Plugin from 'broccoli-plugin';
+import FSTree from 'fs-tree-diff';
+import walkSync from 'walk-sync';
+
+export class Cacher extends Plugin {
+  constructor (inputNodes, options) {
+    super(inputNodes, {
+      ...options,
+      persistentOutput: true,
+    });
+
+    this._previous = [];
+  }
+
+  _hasChanged() {
+    let changed = false;
+    for (let inputPath of this.inputPaths) {
+      const current = FSTree.fromEntries(walkSync.entries(inputPath));
+      const patch = current.calculatePatch(this._previous[inputPath] || []);
+      this._previous[inputPath] = current;
+
+      if (patch.length) {
+        changed = true;
+      }
+    }
+
+    return changed;
+  }
+
+  build() {
+    if (!this._hasChanged()) {
+      return;
+    }
+
+    // do work
+  }
+}
+
+export default function cacher(...params) {
+  return new Cacher(...params);
+}
+```
+
+Let's examine the above. First off we're importing the packages, nothing exciting there.
+In the `constructor()` we're setting `persistentOutput: true` in the options hash to true, this ensures that 
+Broccoli does not delete the output directory before calling `build()`, we need this so our output is preserved 
```

Comment:
The sentence gets pretty long, maybe switch to a period at “`build()`, we”

Reply:
Question I had after reading this: while `persistentOutput: true` preserves the output dir within a broccoli/ember serve, does the directory get preserved between runs of the broccoli/ember serve process?

Reply:
I think it might make sense to mention in a previous section that Broccoli will by default give you an empty output directory (i.e. delete the results of the last invocation of `build()` upon rebuilds); that seems like an important concept to understand before the alternative approach  with `persistentOutput: true` is mentioned in this section.

Reply:
done